ros2_foxglove_bridge/src/message_definition_cache.cpp

```diff
@@ -166,15 +202,53 @@ const MessageSpec& MessageDefinitionCache::load_message_spec(
   }
   const std::string contents{std::istreambuf_iterator(file), {}};
 
-  const MessageSpec& spec =
-    msg_specs_by_definition_identifier_
-      .emplace(definition_identifier,
-               MessageSpec(definition_identifier.format, std::move(contents), package))
-      .first->second;
+  if (is_action_type) {
+    if (definition_identifier.format == MessageDefinitionFormat::MSG) {
+      const auto [goalDef, resultDef, feedbackDef] = split_action_msg_definition(contents);
 
-  // "References and pointers to data stored in the container are only invalidated by erasing that
-  // element, even when the corresponding iterator is invalidated."
-  return spec;
+      // Define type definitions for each action subtype.
+      // These type definitions may include additional fields such as the goal_id.
+      // See also https://design.ros2.org/articles/actions.html
+      const std::map<std::string, std::string> action_type_definitions = {
+        {"_FeedbackMessage", "unique_identifier_msgs/UUID goal_id\n" + feedbackDef},
+        {"_GetResult_Request", "unique_identifier_msgs/UUID goal_id\n"},
+        {"_GetResult_Response", "int8 status\n" + resultDef},
+        {"_SendGoal_Request", "unique_identifier_msgs/UUID goal_id\n" + goalDef},
+        {"_SendGoal_Response", "bool accepted\nbuiltin_interfaces/msg/Time stamp"}};
```

Comment:
The linked design doc discusses these fields at a high level, but where is it documented that these are actually added to the definitions?

Reply:
I also have not found proper documentation for that. Best documentation I could find is the source code that generates the action types:

https://github.com/ros2/rosidl/blob/a57baea5dee6108246db9991e39d28a10a7ee2b0/rosidl_parser/rosidl_parser/definition.py#L720-L727